euphonic/qpoint_phonon_modes.py

```diff
@@ -716,3 +746,171 @@ def from_phonopy(cls: Type[T], path: str = '.',
+            'Md', 'No', 'Lr', 'Rf', 'Db', 'Sg', 'Bh', 'Hs', 'Mt', 'Ds', 'Rg',
+            'Cn', 'Nh', 'Fl', 'Mc', 'Lv', 'Ts', 'Og']
+
+        def get_z(symbol: str) -> int:
+            try:
+                return elements.index(symbol)
+            except ValueError:  # Symbol not found
+                return 0
+
+        def symbols_to_formula(symbols: list[str]) -> str:
+            from collections import Counter
+            symbol_counts = Counter(symbols)
+
+            return "".join(f"{symbol}{symbol_counts[symbol]}"
+                           for symbol in sorted(symbol_counts))
+
+        return dict(
+            natoms = len(crystal.atom_type),
+            lattice = crystal.cell_vectors.to("angstrom").magnitude.tolist(),
+            atom_types = crystal.atom_type.tolist(),
+            atom_numbers = list(map(get_z, crystal.atom_type)),
+            formula = symbols_to_formula(crystal.atom_type),
+            atom_pos_red = crystal.atom_r.tolist(),
+            atom_pos_car = (crystal.atom_r @ crystal.cell_vectors).to("angstrom").magnitude.tolist()
+        )
+
+    @staticmethod
+    def _remove_breaks(distances: np.ndarray, btol: float = 10.) -> list[int]:
+        """Collapse large breaks in cumulative-distance array
+
+        These correspond to discontinuous regions of the x-axis: in euphonic
+        plotting this is usually handled by splitting the spectrum and plotting
+        to new axes, but Phonon Website does not handle this.
+
+        Data is modified in-place
+
+        A list of identified breakpoints is returned
+
+        """
+        diff = np.diff(distances)
+        median = np.median(diff)
+        breakpoints = np.where((diff / median) > btol)[0] + 1
+
+        for breakpoint in reversed(breakpoints):
+            distances[breakpoint:] -= (distances[breakpoint] - distances[breakpoint - 1])
+
+        return breakpoints.tolist()
+
+    @staticmethod
+    def _expand_duplicates(distances: np.ndarray, pad_fraction = 0.001) -> list[int]:
+        diff = np.diff(distances)
+        pad = np.median(diff) * pad_fraction
+
+        duplicates = np.where(diff == 0.)[0] + 1
+        for duplicate in reversed(duplicates):
+            distances[duplicate:] += (distances[duplicate - 1] - distances[duplicate] + pad)
+        return duplicates.tolist()
+
+    @staticmethod
+    def _combine_neighbouring_labels(x_tick_labels: list[tuple[int, str]]
+                                     ) -> list[tuple[int, str]]:
+        """Merge neighbouring labels in x_tick_label list
+
+        If labels are the same, only keep one.
+
+        If labels are different, join with |
+
+        e.g.::
+
+          [(1, "X"), (2, "X"), (4, "A"), (7, "Y"), (8, "Z")]
+
+          -->
+
+          [(1, "X"), (4, "A"), (7, "Y|Z")]
+
+        """
+        labels = dict(x_tick_labels)
+
+        for index in sorted(labels):
+            if index - 1 in labels:
+                if labels.get(index - 1) != labels.get(index):
```

Comment:
Realise what this is doing but the logic here makes me unsure about precisely how it is doing it.

Reply:
Not really sure how to clarify, this looks like over-explaining:
```python
def _combine_neighbouring_labels(x_tick_labels: XTickLabels) -> XTickLabels:
    """Merge neighbouring labels in x_tick_label list

    If labels are the same, only keep one.

    If labels are different, join with |

    e.g.::

      [(1, "X"), (2, "X"), (4, "A"), (7, "Y"), (8, "Z")]

      -->

      [(1, "X"), (4, "A"), (7, "Y|Z")]

    """
    labels = dict(x_tick_labels)

    for index in sorted(labels):
        if index - 1 in labels:
            if labels.get(index - 1) != labels.get(index):
                # Neighbouring labels are different: merge second into first
                labels[index - 1] = f"{labels[index - 1]}|{labels[index]}"
            # Remove second label of pair
            del labels[index]
    return sorted(labels.items())
```

Reply:
I think it's more why 
```python
if index - 1 in labels:
```
 defines them as "the same". That's the logic I can't follow, that only implies the neighbouring point still exists. If the *value* were the same rather than the key, then I could see removing the second label pair. But every time?

Reply:
I'd expect it to produce `[(1, "X|Y"), (3, "X")]` which isn't brilliant but is ok given that this was a terrible input.

Reply:
I'll see if I can rewrite more obviously using `pairwise`

Reply:
Ok, so that isn't what I expected. I was thinking it would leave things unchanged. But I guess now that I look again, it makes more sense. 

I think it's actually ok, but I think I would add a third part to the `if`: neighbours (same), neighbours (different), non-neighbours.

Reply:
How about this, which does a slightly different but acceptable thing `[(3, "X|Y|X")]` and might be more readable:

```python
def _combine_neighbouring_labels(x_tick_labels: XTickLabels) -> XTickLabels:
    """Merge neighbouring labels in x_tick_label list

    If labels are the same, only keep one.

    If labels are different, join with |

    e.g.::

      [(1, "X"), (2, "X"), (4, "A"), (7, "Y"), (8, "Z")]

      -->

      [(1, "X"), (4, "A"), (7, "Y|Z")]

    """
    labels = dict(x_tick_labels)

    for (index, next_index) in pairwise(sorted(labels)):

        if index + 1 == next_index:  # Labels are neighbours

            if labels[index] != labels[next_index]:  # Combine differing labels
                labels[next_index] = f"{labels[index]}|{labels[next_index]}"

            del labels[index]  # Drop redundant label

    return sorted(labels.items())

```

Reply:
Much clearer to me about the intent and the function.